src/Security/OpenIdConfigurationProviderManager.php

```diff
@@ -0,0 +1,107 @@
+use ItkDev\OpenIdConnect\Exception\ItkOpenIdConnectException;
+use ItkDev\OpenIdConnect\Exception\JsonException;
+use ItkDev\OpenIdConnect\Exception\KeyException;
+use ItkDev\OpenIdConnect\Exception\MissingParameterException;
+use ItkDev\OpenIdConnect\Exception\NegativeLeewayException;
+use ItkDev\OpenIdConnect\Exception\ValidationException;
+use ItkDev\OpenIdConnect\Security\OpenIdConfigurationProvider;
+use ItkDev\OpenIdConnectBundle\Exception\InvalidProviderException;
+use League\OAuth2\Client\Provider\AbstractProvider;
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use League\OAuth2\Client\Provider\GenericResourceOwner;
+use League\OAuth2\Client\Token\AccessToken;
+use League\OAuth2\Client\Tool\RequestFactory;
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Cache\InvalidArgumentException;
+use Psr\Http\Message\ResponseInterface;
+use RobRichards\XMLSecLibs\XMLSecurityKey;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\Routing\RouterInterface;
+
+class OpenIdConfigurationProviderManager
+{
+    /**
+     * @var array
+     */
+    private $config;
+
+    /**
+     * @var array
+     */
+    private $providers;
+
+    /**
+     * @var RouterInterface
+     */
+    private $router;
+
+    public function __construct(RouterInterface $router, array $config)
+    {
+        $this->router = $router;
+        $this->config = $config;
+    }
+
+    /**
+     * Get all providers.
+     *
+     * @return array|OpenIdConfigurationProvider[]
+     * @throws ItkOpenIdConnectException
+     */
+    public function getProviders(): array
+    {
+        if (null === $this->providers) {
+            $this->providers = [];
+            foreach ($this->config['providers'] as $key => $options) {
+                $providerOptions = [
+                        'openIDConnectMetadataUrl' => $options['metadata_url'],
+                        'clientId' => $options['client_id'],
+                        'clientSecret' => $options['client_secret'],
+                    ] + $this->config['default_providers_options'];
+
+                if (isset($options['redirect_uri'])) {
+                    $providerOptions['redirectUri'] = $options['redirect_uri'];
+                } elseif (isset($options['redirect_route'])) {
+                    $providerOptions['redirectUri'] = $this->router->generate(
+                        $options['redirect_route'],
+                        $options['redirect_route_parameters'] ?? [],
+                        RouterInterface::ABSOLUTE_URL
+                    );
+                }
+
+                $this->providers[$key] = new OpenIdConfigurationProvider($providerOptions);
+            }
+        }
+
+        return $this->providers;
+    }
+
+    /**
+     * Get a provider by key.
+     *
+     * @param string $key
+     * @return OpenIdConfigurationProvider
+     * @throws InvalidProviderException
+     */
+    public function getProvider(string $key): OpenIdConfigurationProvider
```

Comment:
Suggestion: refactor this so that only the requested provider gets instantiated. Theoretically a given request will only ever need one provider, but we instantiate all configured providers, then only provide one. Not the most expensive operation to waste but still a nice optimization.

Reply:
Fixed in https://github.com/itk-dev/openid-connect-bundle/pull/11/commits/87d6d846d93c5ae9f7f2161540b49c39d3841a48